pkg/yurthub/storage/disk/storage.go

```diff
@@ -343,6 +343,84 @@ func (ds *diskStorage) Update(key string, contents []byte) error {
 	return os.Rename(filepath.Join(ds.baseDir, tmpKey), filepath.Join(ds.baseDir, key))
 }
 
+// Replace will delete all files under rootKey dir and create new files with contents.
+func (ds *diskStorage) Replace(rootKey string, contents map[string][]byte) error {
+	if rootKey == "" {
+		return storage.ErrKeyIsEmpty
+	} else if len(contents) == 0 {
+		return storage.ErrKeyHasNoContent
+	}
+
+	for key := range contents {
+		if !strings.Contains(key, rootKey) {
+			return storage.ErrRootKeyInvalid
+		}
+	}
+
+	if !ds.lockKey(rootKey) {
+		return storage.ErrStorageAccessConflict
+	}
+	defer ds.unLockKey(rootKey)
+
+	// 1. mv old dir into tmp_dir when rootKey dir already exists
+	absPath := filepath.Join(ds.baseDir, rootKey)
+	tmpRootKey := getTmpKey(rootKey)
+	tmpPath := filepath.Join(ds.baseDir, tmpRootKey)
+	dirExisted := false
+	if info, err := os.Stat(absPath); err == nil {
+		if info.IsDir() {
+			err := os.Rename(absPath, tmpPath)
+			if err != nil {
+				return err
+			}
+			dirExisted = true
+		}
+	}
+
+	// 2. create new file with contents
+	for key, data := range contents {
+		err := ds.create(key, data)
+		if err != nil {
+			klog.Errorf("failed to create %s in replace, %v", key, err)
```

Comment:
If error happens, do we have retry mechanism? If not, do we have mechanism to do consistency check?

Reply:
I have not seriously thought about how to guarantee that object can be cached successfully. so i'd like to keep this subject as a TODO item. and we can solve it in the soon future.